rascal-lsp/src/main/rascal/lang/rascal/tests/rename/TestUtils.rsc

```diff
@@ -0,0 +1,288 @@
+@license{
+Copyright (c) 2018-2023, NWO-I CWI and Swat.engineering
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+
+1. Redistributions of source code must retain the above copyright notice,
+this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+this list of conditions and the following disclaimer in the documentation
+and/or other materials provided with the distribution.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+}
+@bootstrapParser
+module lang::rascal::tests::rename::TestUtils
+
+import lang::rascal::lsp::refactor::Rename; // Module under test
+
+import IO;
+import List;
+import Location;
+import Set;
+import String;
+
+import lang::rascal::\syntax::Rascal; // `Name`
+
+import lang::rascalcore::check::Checker;
+import lang::rascalcore::check::BasicRascalConfig;
+import lang::rascalcore::check::RascalConfig;
+import lang::rascalcore::compile::util::Names;
+
+import analysis::diff::edits::ExecuteTextEdits;
+
+import util::FileSystem;
+import util::Math;
+import util::Reflective;
+
+
+//// Fixtures and utility functions
+data TestModule = byText(str name, str body, set[int] expectedRenameOccs, str newName = name)
+                | byLoc(loc file, set[int] expectedRenameOccs, str newName = name);
+
+private list[DocumentEdit] sortEdits(list[DocumentEdit] edits) = [sortChanges(e) | e <- edits];
+
+private DocumentEdit sortChanges(changed(loc l, list[TextEdit] edits)) = changed(l, sort(edits, bool(TextEdit e1, TextEdit e2) {
+    return e1.range.offset < e2.range.offset;
+}));
+private default DocumentEdit sortChanges(DocumentEdit e) = e;
+
+private void verifyTypeCorrectRenaming(loc root, list[DocumentEdit] edits, PathConfig pcfg) {
+    executeDocumentEdits(sortEdits(edits));
+    remove(pcfg.resources);
+    RascalCompilerConfig ccfg = rascalCompilerConfig(pcfg)[forceCompilationTopModule = true][verbose = false][logPathConfig = false];
+    throwAnyErrors(checkAll(root, ccfg));
+}
+
+bool ASSERT_EQ(&T expected, &T actual) {
```

Comment:
I'm not too fond of the SCREAMY NAME here. And it is really an "assert"?

Reply:
It's and assert in the context of testing; it fails fast, like GTest's assert (vs expect, which first collects all failures and reports them all at once).
This started out as a little helper to investigate failing tests and I think something like this would be a useful addition to the stdlib. I would be happy adding that there if we can agree on a name. `expect_eq` or `assert_eq` would be fine with me as well, for example.